Wrappers/Python/cil/optimisation/algorithms/SPDHG.py

```diff
@@ -124,100 +247,118 @@ def set_up(self, f, g, operator, tau=None, sigma=None, \
+                          for i in range(self.ndual_subsets)]
+        else:
+            if len(precalculated_norms) == self.ndual_subsets:
+                if all(isinstance(x, Number) for x in precalculated_norms):
+                    if all(x > 0 for x in precalculated_norms):
+                        pass
+                    else:
+                        raise ValueError(
+                            "The norms of the operators should be positive")
+                else:
+                    raise ValueError(
+                        "The norms of the operators should be a Number")
+            else:
+                raise ValueError(
+                    "Please pass a list of floats to the precalculated norms with the same number of entries as number of operators")
+            self.norms = precalculated_norms
+
+        if sampler is None:
+            if self.prob_weights is None:
+                self.prob_weights = [1/self.ndual_subsets] * self.ndual_subsets
+            self.sampler = Sampler.randomWithReplacement(
+                self.ndual_subsets,  prob=self.prob_weights)
+        else:
+            if not isinstance(sampler, Sampler):
+                raise ValueError(
+                    "The sampler should be an instance of the CIL Sampler class")
+            self.sampler = sampler
+            if sampler.prob is None:
+                self.prob_weights = [1/self.ndual_subsets] * self.ndual_subsets
+            else:
+                self.prob_weights = sampler.prob
+
+        # might not want to do this until it is called (if computationally expensive)
+        self.set_step_sizes_custom()
+
+        # initialize primal variable
         if initial is None:
             self.x = self.operator.domain_geometry().allocate(0)
         else:
             self.x = initial.copy()
-        
+
         self.x_tmp = self.operator.domain_geometry().allocate(0)
-        
+
         # initialize dual variable to 0
         self.y_old = operator.range_geometry().allocate(0)
-        
+
         # initialize variable z corresponding to back-projected dual variable
         self.z = operator.domain_geometry().allocate(0)
-        self.zbar= operator.domain_geometry().allocate(0)
+        self.zbar = operator.domain_geometry().allocate(0)
         # relaxation parameter
         self.theta = 1
         self.configured = True
         logging.info("{} configured".format(self.__class__.__name__, ))
-        
+
     def update(self):
         # Gradient descent for the primal variable
         # x_tmp = x - tau * zbar
-        self.x.sapyb(1., self.zbar,  -self.tau, out=self.x_tmp)
-            
-        self.g.proximal(self.x_tmp, self.tau, out=self.x)
-        
+        self.x.sapyb(1., self.zbar,  -self._tau, out=self.x_tmp)
+
+        self.g.proximal(self.x_tmp, self._tau, out=self.x)
+
         # Choose subset
-        i = int(np.random.choice(len(self.sigma), 1, p=self.prob))
-        
+        i = self.sampler.next()
```

Comment:
Or

```suggestion
        i = next(self.sampler)
```

I don't know which is better